[WIP] Parameter optimization

selenasingh opened the pull request:
Created to receive feedback

selenasingh committed:
modify shell scripts

'nrnivmodl' creates a directory called 'arm64' on Macs, rather than 'x86_64'. Shell scripts were modified here to accommodate for this difference.

selenasingh committed:
preventing 'return 0' error

The 'return 0' within 'initial' blocks causes issues for the C compiler, which prevents the entire network from running. Commenting out the verbatim blocks prevents this issue.

selenasingh committed:
Revert "modify shell scripts"

This reverts commit 3852f6e17692c3652c324b78d70a39a7e6f3e4ff.

selenasingh committed:
modify shell scripts

'nrnivmodl' creates a directory called 'arm64' on Macs, rather than 'x86_64'. Shell scripts were modified here to accommodate for this difference.

selenasingh committed:
no change made here, in conflict

selenasingh committed:
creating IV curves

selenasingh committed:
updating VClamp docstring

selenasingh committed:
Update optimize_params.py

selenasingh committed:
able to tune single gc to 1 freq

abrahamnunes commented on optimize_params.py:
```diff
@@ -0,0 +1,262 @@
+- figure out how to deal with the parameters in a nice way
+"""
+
+#imports
```
No need to comment this :) (just a sticky point, but to make code shorter)

selenasingh replied:
I've always used Spyder as my IDE, and it has always added this little comment (Created on blahblah, author selenasingh, etc) at the beginning of each new file, which I've kind of just accepted and now use to add notes to myself. Will remove for future things :)

abrahamnunes commented on optimize_params.py:
```diff
@@ -0,0 +1,262 @@
+import pandas as pd
+import matplotlib.pyplot as plt
+import pylab              
+from random import Random # pseudorandom number generation
```
Why not use `np.random.RandomState`? Then you can declare a distinct generator for each object that requires one

selenasingh replied:
Will try this :) thank you! I used the same random number generator from the tutorial file for now.

abrahamnunes commented on optimize_params.py:
```diff
@@ -0,0 +1,262 @@
+class optimizeparams(object):
+    def __init__(self, 
+                 cell,
+                 num_inputs = 18, #number of parameters to be optimized
```
I would suggest doing this a bit differently, and it should help to get rid of a lot of the copied/pasted code below. Suggested strategy: 

Maybe change `num_inputs` to `free_params`, and have `free_params` accept a dictionary of the following form: `free_params={'mech': ['param1', 'param2']}`. More specifically for your case, 

``` python
free_params = {
    'HT' : ['gbar'],
    'LT' : ['gbar'], 
    'bk' : ['gkbar'], 
    'ichan2' : ['gnatbar', 'gkfbar', 'gksbar', 'gl', 'ggabaa'],
    'ka' : ['gkabar'],
    'kir' : ['gkbar','kl','at','bt'],
    'km' : ['gbar'], 
    'lca' : ['glcabar'], 
    'nca' : ['gncabar'],
    'sk' : ['gskbar'], 
    'tca' : ['gcatbar']
}
```

Then in the `retrieve_baseline_params()` method, you can do the following: 

``` python 
baseline = [] 
for key in self.free_params.keys():
    for val in self.free_params[key]: 
        baseline.append(self.cell_dict['secs']['soma']['mechs'][key][val])

self.num_inputs = len(baseline)

return baseline
```

abrahamnunes replied:
A similar approach could be used in your method `build_optimizedcell()`

selenasingh replied:
Thank you so much for this very helpful suggestion!! I know how I've dealt with the parameters is not ideal, and knew that it would take a while for me to figure out (not v familiar with dictionaries), so I thought I would focus on getting it working first and then make improvements. Will try out your suggestion next :)

abrahamnunes commented on optimize_params.py:
```diff
@@ -0,0 +1,262 @@
+
+#----------run simulations 
+init = optimizeparams(gc)
+baselineparams = init.retrieve_baseline_params()
```
why not store the baseline parameter values as an attribute within the object? It could avoid this additional function call (and make the code more readable). 

In general, if you can hide boilerplate and data admin stuff inside of an object method, your code will be more readable (at least if you are doing object oriented programming, as you are here)

selenasingh replied:
Will do :) I agree that would be better. A large part of this piece was to confirm that the functions were actually working and returning what I hoped they would.

abrahamnunes commented on optimize_params.py:
```diff
@@ -0,0 +1,262 @@
+  
+
+#----------run simulations 
+init = optimizeparams(gc)
```
avoid calling things`init` if possible.

selenasingh replied:
that was a bad choice, I agree. Will be more mindful next time.

abrahamnunes commented on optimize_params.py:
```diff
@@ -0,0 +1,262 @@
+baselinecell = init.curr_inj(0.33)
+
+op = optimizeparams(gc)
+newparams = op.find_bestcandidate()
```
you could just store this in the object as an attribute no? Will make the code shorter.

selenasingh replied:
Will implement this change

abrahamnunes commented on optimize_params.py:
```diff
@@ -0,0 +1,262 @@
+
+op = optimizeparams(gc)
+newparams = op.find_bestcandidate()
+newcell = op.build_optimizedcell()
```
As above

abrahamnunes commented on optimize_params.py:
```diff
@@ -0,0 +1,262 @@
+newparams = op.find_bestcandidate()
+newcell = op.build_optimizedcell()
+
+#----------CONSTRUCT PARAM COMPARISON TABLE 
```
Would be good to just write a function that can do this as part of your object above, or which accepts the object and spits out a table. This is because the paramnames and values will already be stored in the object so you can avoid redefining them here.

selenasingh replied:
This is a great suggestion for moving forward. The table was written up immediately before sharing with you because I realized I should probably demonstrate that the parameters have actually changed post optimization (I was using Spyder's variable explorer to confirm prior to this addition).

abrahamnunes commented on optimize_params.py:
```diff
@@ -0,0 +1,262 @@
+
+
+#----------PLOTTING
+figure, axis = plt.subplots(2,1, constrained_layout=True)
```
best to save plots to file (in the `figures` directory, then an appropriate subdirectory), and set it so that mpl does not show the figures by default. I think it is something like `mpl.Agg('off')` or something like that... This will allow the scripts to be run remotely via SSH without a gui. If you dont do this and you try to run remotely, it will crash.

selenasingh replied:
Making this more SSH-friendly was also on my list--I'll take a look at this `mpl` situation :) thank you!

abrahamnunes commented on run.sh:
```diff
@@ -13,7 +13,7 @@ mkdir figures/fi-curves
 # Run DG Model 
 nrnivmodl mods
-x86_64/special main.hoc
+nrniv main.hoc
```
Curious about why this change was made. Do you know something I dont? :)

selenasingh replied:
`nrnivmodl` creates a folder called `arm64` (instead of `x86_64`) on macs. `nrniv` will link to `special` by itself to incorporate the mod object files without needing to specify the folder. I thought you needed to specify `special` if you're using CoreNEURON (I think I was struggling with the installation a while ago for this).

abrahamnunes commented on clamps.py:
```diff
@@ -0,0 +1,183 @@
+# ========================================================================
```
Not sure why this is showing up totally new here. Were any changes made to the original?

selenasingh replied:
I'm not sure either, I didn't make any changes recently.. however a few commits ago I changed the docstring for `VClamp`. I don't really understand why it's showing up as new this time.

abrahamnunes commented on IVdata.py:
```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python3
```
I presume this is just like a testing or "sandbox" script?

selenasingh replied:
I wrote this with the intention to ultimately use it to tune the cells to specific I-V curves, but yes, it was also an exercise in practicing how to use `clamps`. Will start with FI curves, but if we want to use the current data we'll need something that creates those IV curves as well :)

abrahamnunes reviewed (commented):
Some initial comments :) Great work so far!

selenasingh reviewed (commented):
Replies to all suggestions :)

selenasingh committed:
implementing AN suggested changes

selenasingh committed:
code for extracting avg FI from .csv

selenasingh committed:
re-name find-rheobase to find_rheobase

will make importing easier

selenasingh committed:
modify run.sh to work w apple m1

selenasingh committed:
Update .gitignore

selenasingh committed:
Update run.sh

selenasingh committed:
opt params to fit to avg FI curve

selenasingh committed:
tidying up

selenasingh committed:
exploring similarity measures

selenasingh committed:
adding LR and NR csv/dfs

selenasingh committed:
implementing MSE as fitness eval

selenasingh committed:
better parameter choices

selenasingh committed:
Adding SEM calculation

selenasingh committed:
temporarily commenting out one line

selenasingh committed:
avg simulated FIs

plot avg simulated FIs against avg FI from data, change parameter choices, nicer figures, export optimized params as csv.

selenasingh committed:
Mult neurons, add bkg noise, add params to shift activation/inactivation curves.

abrahamnunes commented on optimize_params.py:
```diff
@@ -0,0 +1,335 @@
+rawnr = pd.read_csv("rawdata/NR_FI.csv")
+
+
+class optimizeparams(object):
```
would be good to have a docstring describing attributes, methods, datatypes and the like.

selenasingh replied:
Sounds good, I've implemented this.

abrahamnunes commented on optimize_params.py:
```diff
@@ -0,0 +1,335 @@
+        self.noiselvl = random.uniform(0.8, 1)
+        self.weight = 0.0001  # random.uniform(0.01, 0.1)
+
+    def curr_inj(self, current, delay=0, duration=1000, noise=False):
```
then at the methods it would be good to have a docstring describing what is happening, arguments and their data types, return types etc.

selenasingh replied:
See above.

abrahamnunes commented on optimize_params.py:
```diff
@@ -0,0 +1,335 @@
+            iclamp = ICNoise(self.cell_dict, delay=delay, duration=duration, T=duration + delay * 2)
+            res = iclamp(current, self.noiselvl, self.weight)
+        else:
+            iclamp = IClamp(self.cell_dict, delay=delay, duration=duration, T=duration + delay * 2)
```
In the future we might consider rewriting `IClamp` and `VClamp` to not have to re-initialize the whole thing for these optimizations. However, here it is not a huge priority.

abrahamnunes commented on optimize_params.py:
```diff
@@ -0,0 +1,335 @@
+
+    def find_bestcandidate(self):
+        rand = Random()
+        rand.seed()  # will take current time as seed
```
This will not be reproducible then... best to set a seed...

also I'm curious about your choice to use Random() rather than `numpy.random.RandomState`. why the former?

selenasingh replied:
I've changed this so it's reproducible now, thank you! I've tried using `np.random.RandomState` and it's not compatible with some of the inspyred functions. I can write my own versions of those functions so that they work with `np.random.RandomState` if you'd like, but for now I've left this as `Random`.

abrahamnunes commented on optimize_params.py:
```diff
@@ -0,0 +1,335 @@
+    def evaluate_netparams(self, candidates, args):
+        self.fitnessCandidates = []
+
+        for icand, cand in enumerate(candidates):
```
you have a decent solution below, I think... something along the lines of 

```
i = 0
for k in free_params.keys():
    for v in free_params[k]:
         self.cell_dict['secs']['soma']['mechs'][k][v] = cand[i]
         i += 1
```

...unless I'm missing something here?

abrahamnunes replied:
also not sure why you are using `enumerate` when `icand` doesn't appear to be getting called anywhere. you could just do

```
for cand in candidates: 
     i = 0
     for k in free_params.keys():
         for v in free_params[k]:
             self.cell_dict['secs']['soma']['mechs'][k][v] = cand[i]
             i += 1
```

selenasingh replied:
Thank you! I've made this change. I thought that `icand` was used by one of the inspyred functions, but I was wrong. There used to be a print statement that referred to `icand` that I removed a while ago and forgot about. This solution works beautifully.

abrahamnunes commented on optimize_params.py:
```diff
@@ -0,0 +1,335 @@
+        plt.close(fig2)
+
+
+gc = importgc()  # re-loading to ensure baseline params are the same each time
```
This could all be shortened into a couple of lines...

```
opt_results = [
    optimizeparams(importgc(), free_params, rawnrn, group, condition, "MSE")
    for (rawnrn, group) in [(rawhc, "HC"), (rawlr, "LR"), (rawnr, "NR")] 
    for condition in ["CTRL", "LITM"]
]

for r in opt_results: 
    r.return_summarydata() 
```

You might also consider making "MSE" a default value for `optimizeparams`, since you are calling it every time as the standard. 

`op_x.return_summarydata()` might also be run automatically, since it looks like you are calling it every time. 

finally, `importgc()` looks like it should be called within `optimizeparams`, since calling it outside doesn't really seem to add much... unless I'm missing some recursion somewhere?

selenasingh replied:
1.  I've made `MSE` the standard option for curve comparison, and have also implemented your suggested change to shorten the code at the bottom of the file, thank you!
2. I've included `return_summarydata()` , re-named to `plot_results()` in `__init__` so now that method gets called when you instantiate the class. 
3. `importgc()` functions to ensure that each population/condition combination starts with the same base parameters (I have this here as a temporary solution). Otherwise, the optimized parameters from the previous condition end up being used as the baseline for the next condition. `revert_to_baseline` is supposed to revert parameters back to the baseline gc, but it doesn't seem to work. I'm still working on figuring out what's wrong with it.

abrahamnunes commented on clamps_noise.py:
```diff
@@ -0,0 +1,103 @@
+import numpy as np
```
consider merging this file with `clamps.py` and simply adding noise as an optional setting. otherwise remove the file, I would say..

good rule of thumb is that any time you find yourself copying and pasting code, that is a good time to abstract your code into something more general to encompass all the intended cases

selenasingh replied:
I've merged those two files and have included noise as an option :)

abrahamnunes commented on cleanup.sh:
```diff
@@ -1,3 +1,3 @@
-rm -R x86_64/
+rm -R arm64/
```
rather than changing the whole file, it would be best to have a different file called run_m1.sh, and cleanup_m1.sh, that way we can still run on linux machines if necessary without having to change the code again

selenasingh replied:
See below

abrahamnunes commented on run.sh:
```diff
@@ -2,20 +2,30 @@
+#if wanting to run natively on m1
 nrnivmodl mods
-x86_64/special main.hoc
+#nrniv main.hoc
```
shouldn't this be after the python code? (once it is uncommented)

abrahamnunes replied:
also please see comment on having separate run and cleanup files for m1, so that we can still use these original ones on linux without having to modify again.

selenasingh replied:
1. Thank you for catching that. I've switched the order now. 
2. I've changed the file names to have `_m1`, should I include separate `run.sh` and `cleanup.sh` meant for linux as well on this branch?

selenasingh committed:
address AN comments, incorporate volt-clamp data

selenasingh committed:
remove cshift params, improve min/max bounds

selenasingh committed:
best min/max bounds, good optimization performance

selenasingh committed:
incorporate params into network

selenasingh committed:
Fixing MCs, prep for network E/I re-balance.

selenasingh committed:
shift each FI, manual add AN's spont. activity.

selenasingh committed:
optimize dendrite parameters

selenasingh committed:
plot summary figure

selenasingh committed:
dend min/max bounds. aggregate plot.

selenasingh committed:
good opt'm, prep for PS exps

selenasingh committed:
DG updates, run scripts, negative IV curves, analysis update

selenasingh committed:
WTA analysis
